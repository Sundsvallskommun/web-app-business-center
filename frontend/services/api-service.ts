import Router from 'next/router';
import { apiURL } from '@utils/api-url';
import axios from 'axios';
import { ApiResponseMeta } from '@interfaces/service';

export interface Data {
  error?: string;
  _meta?: ApiResponseMeta;
}

export interface ApiResponse<T> {
  data: T;
  message: string;
}

let isChangingRoute = false;
let isRedirectingToLogin = false;

Router.events.on('routeChangeStart', (route) => {
  isChangingRoute = true;
});

Router.events.on('routeChangeComplete', (route) => {
  isRedirectingToLogin = false;
  isChangingRoute = false;
  if (route === '/login') {
    isRedirectingToLogin = false;
  }
});

export const handleError = (error) => {
  /*let s = '';
  if (error.response) {
    // The request was made and the server responded with a status code
    // that falls out of the range of 2xx
    s += `Server responded with ${error.response.status} ${error.response.data.message}`;
  } else if (error.request) {
    // The request was made but no response was received
    // `error.request` is an instance of XMLHttpRequest in the browser and an instance of
    // http.ClientRequest in node.js
    s += `Server did not respond`;
  } else {
    // Something happened in setting up the request that triggered an Error
    s += `Unknown error: ${error.message}`;
  }
  s += ` for url ${error.config.url}`;
  */
  // console.error(s);

  // FIXME: we should have a logged in guard?
  if (error.response.status === 401 && Router.pathname !== '/login' && !isRedirectingToLogin) {
    isRedirectingToLogin = true;
    Router.push('/login');
  }

  throw error;
};

const options = {
  headers: {
    'Content-Type': 'application/json',
  },
  withCredentials: true,
};

const get = <T>(url: string) => axios.get<T>(apiURL(url), options).catch(handleError);

const post = <T>(url: string, data: any) => {
  return axios.post<T>(apiURL(url), data, options).catch(handleError);
};

const remove = <T>(url: string) => {
  return axios.delete<T>(apiURL(url), options).catch(handleError);
};

const patch = <T>(url: string, data: any) => {
  return axios.patch<T>(apiURL(url), data, options).catch(handleError);
};

export const apiService = { get, post, patch, delete: remove };
