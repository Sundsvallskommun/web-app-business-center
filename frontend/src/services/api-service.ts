import { __DEV__ } from '@sk-web-gui/react';
import {
  DefaultError,
  QueryClient,
  QueryKey,
  UseMutationOptions,
  UseMutationResult,
  UseQueryOptions,
  UseQueryResult,
  useMutation,
  useQuery,
} from '@tanstack/react-query';
import { apiURL } from '@utils/api-url';
import axios, { AxiosError, AxiosRequestConfig, AxiosResponse } from 'axios';
import { create } from 'zustand';
import { devtools } from 'zustand/middleware';

export interface ApiResponse<T> {
  data: T;
  message: string;
}

export const handleError = (error) => {
  if (error?.response?.status === 401 && !window?.location.pathname.includes('login')) {
    window.location.href = `/login?path=${window.location.pathname}&failMessage=${error.response.data.message}`;
  }
};

const defaultOptions = {
  headers: {
    'Content-Type': 'application/json',
  },
  withCredentials: true,
};

const get = <T>(url: string, options?: { [key: string]: unknown }) =>
  axios.get<T>(apiURL(url), { ...defaultOptions, ...options });

const post = <T>(url: string, data: unknown = undefined, options?: { [key: string]: unknown }) => {
  return axios.post<T>(apiURL(url), data, { ...defaultOptions, ...options });
};

const remove = <T>(url: string, options?: { [key: string]: unknown }) => {
  return axios.delete<T>(apiURL(url), { ...defaultOptions, ...options });
};

const patch = <T>(url: string, data: unknown = undefined, options?: { [key: string]: unknown }) => {
  return axios.patch<T>(apiURL(url), data, { ...defaultOptions, ...options });
};

const put = <T>(url: string, data: unknown = undefined, options?: { [key: string]: unknown }) => {
  return axios.put<T>(apiURL(url), data, { ...defaultOptions, ...options });
};

export const apiService = { get, post, put, patch, delete: remove };

export const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      refetchOnWindowFocus: false, // default: true
      staleTime: 1000 * 60 * 5, // 5 minutes
      retry: (failureCount, error: AxiosError) => {
        console.log(`Response Code: ${error.response?.status} failureCount: ${failureCount}`);
        const shouldRetry = error.response?.status === 500 && failureCount < 3;
        if (shouldRetry) console.log('Retrying ....!');
        // retry on 500 errors for max 3 times
        return shouldRetry;
      },
    },
  },
});

interface State {
  queryClient: QueryClient;
}

const initialState: State = {
  queryClient: queryClient,
};

export const useApiService = create<State>()(
  devtools(
    () => ({
      ...initialState,
    }),
    { name: 'api-service', enabled: __DEV__ }
  )
);
type Method = 'get' | 'post' | 'put' | 'patch' | 'delete';

type ContextDefault = { newBody: UseApiProps['body'] };

interface UseApiQueryProps<
  TQueryFnData = unknown,
  TError = Error,
  TData = TQueryFnData,
  TQueryKey extends QueryKey = QueryKey,
  TContext = ContextDefault,
> {
  url: string;
  method: 'get';
  axiosParameters?: AxiosRequestConfig;
  dataHandler?: (data: TQueryFnData) => TData;
  body?: Record<string, unknown>;
  queryKey?: TQueryKey;
  queryOptions?: UseQueryOptions<TQueryFnData, TError, TData, TQueryKey>;
  mutationOptions?: UseMutationOptions<TData, TError, unknown, TContext>;
}

interface UseApiMutationProps<
  TQueryFnData = unknown,
  TError = Error,
  TData = TQueryFnData,
  TQueryKey extends QueryKey = QueryKey,
  TContext = ContextDefault,
> {
  url: string;
  method: 'post' | 'put' | 'patch' | 'delete';
  axiosParameters?: AxiosRequestConfig;
  dataHandler?: (data: TQueryFnData) => TData;
  body?: Record<string, unknown>;
  queryKey?: TQueryKey;
  queryOptions?: UseQueryOptions<TQueryFnData, TError, TData, TQueryKey>;
  mutationOptions?: UseMutationOptions<TData, TError, unknown, TContext>;
}

type UseApiProps<
  TQueryFnData = unknown,
  TError = Error,
  TData = TQueryFnData,
  TQueryKey extends QueryKey = QueryKey,
  TMethod extends Method = 'get',
  TContext = ContextDefault,
> = TMethod extends 'get'
  ? UseApiQueryProps<TQueryFnData, TError, TData, TQueryKey, TContext>
  : UseApiMutationProps<TQueryFnData, TError, TData, TQueryKey, TContext>;

export const defaultApiCall: <TQueryFnData = unknown>(config: {
  url: string;
  method?: Method;
  body?: unknown;
  axiosParameters?: AxiosRequestConfig;
  context?: {
    queryKey: QueryKey;
    signal: AbortSignal;
    meta: Record<string, unknown> | undefined;
    pageParam?: unknown;
    direction?: unknown;
  };
}) => Promise<AxiosResponse<ApiResponse<TQueryFnData>>> = <TQueryFnData>(config) => {
  return apiService[config.method ?? 'get']<TQueryFnData>(config.url, config.body, {
    signal: config.context?.signal,
    ...config.axiosParameters,
    params: { pageParam: config.context?.pageParam, meta: config.context?.meta, ...config.axiosParameters?.params },
  });
};

type UseApiResult<
  TQueryFnData = unknown,
  TError = Error,
  TData = TQueryFnData,
  TMethod extends Method = 'get',
  TContext = ContextDefault,
> = TMethod extends 'get' ? UseQueryResult<TData, TError> : UseMutationResult<TData, TError, unknown, TContext>;

// Overloads for better type inference
export function useApi<
  TQueryFnData = unknown,
  TError = Error,
  TData = TQueryFnData,
  TQueryKey extends QueryKey = [''],
  TContext = ContextDefault,
>(
  props: UseApiQueryProps<TQueryFnData, TError, TData, TQueryKey, TContext>,
  queryClient?: QueryClient
): UseQueryResult<TData, TError>;

export function useApi<
  TQueryFnData = unknown,
  TError = Error,
  TData = TQueryFnData,
  TQueryKey extends QueryKey = [''],
  TContext = ContextDefault,
>(
  props: UseApiMutationProps<TQueryFnData, TError, TData, TQueryKey, TContext>,
  queryClient?: QueryClient
): UseMutationResult<TData & { error?: DefaultError }, TError, unknown, TContext>;

export function useApi<
  TQueryFnData = unknown,
  TError = Error,
  TData = TQueryFnData,
  TQueryKey extends QueryKey & string[] = [''],
  TMethod extends Method = 'get',
  TContext = ContextDefault,
>(
  props: UseApiProps<TQueryFnData, TError, TData, TQueryKey, TMethod>,
  queryClient?: QueryClient
): UseApiResult<TQueryFnData, TError, TData, TMethod, TContext> {
  const {
    url,
    method,
    axiosParameters,
    dataHandler = (data) => data,
    body,
    queryKey = [url] as TQueryKey,
    queryOptions,
    mutationOptions,
  } = props;
  const _store_queryClient = useApiService((s) => s.queryClient);
  const _queryClient = queryClient ?? _store_queryClient;

  const defaultQueryCall = async (context) =>
    defaultApiCall<TQueryFnData>({ url, method, body, axiosParameters, context }).then((res) =>
      dataHandler(res.data.data)
    );

  const defaultMutationCall = async (body: UseApiProps['body']) => {
    try {
      return defaultApiCall<TQueryFnData>({ url, method, body, axiosParameters }).then((res) =>
        dataHandler(res.data.data)
      );
    } catch (error) {
      handleError(error);
      return { error };
    }
  };

  if (method === 'get') {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    return useQuery(
      {
        queryKey,
        queryFn: defaultQueryCall,
        throwOnError: (error) => {
          handleError(error);
          return false;
        },
        ...queryOptions,
      },
      _queryClient
    ) as UseApiResult<TQueryFnData, TError, TData, TMethod, TContext>;
  } else {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    return useMutation(
      {
        mutationFn: defaultMutationCall,
        onMutate: async (body: Record<string, unknown>) => {
          await _queryClient.cancelQueries({ queryKey });
          const newBody = { ...body };
          return { newBody };
        },
        onSuccess: (result) => {
          _queryClient.setQueryData<TQueryFnData & { error?: DefaultError }>(queryKey, result);
        },
        throwOnError: (error) => {
          handleError(error);
          return false;
        },
        ...mutationOptions,
      },
      _queryClient
    ) as UseApiResult<TQueryFnData, TError, TData, TMethod, TContext>;
  }
}
