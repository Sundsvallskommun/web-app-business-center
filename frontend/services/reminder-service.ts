import dayjs from 'dayjs';
import { statusCodes } from '../interfaces/status-codes';
import { ApiResponse, apiService, Data } from './api-service';

interface ReminderResponse {
  reminderId: string;
  personId: string;
  action: string;
  note: string;
  caseId: string;
  caseLink: string;
  reminderDate: string;
  caseType: string;
  externalCaseId: string;
  created: string;
  createdBy: string;
  modified: string;
  modifiedBy: string;
}

export type SendNoteOmittedTypes = 'reminderId' | 'personId' | 'created' | 'createdBy' | 'modified' | 'modifiedBy';
export type SendReminder = Omit<ReminderResponse, SendNoteOmittedTypes>;

export interface ReminderLabel {
  label: string;
  sortable: boolean;
  shownForStatus: statusCodes;
  screenReaderOnly: boolean;
}

export interface RemindersData extends Data {
  reminders: ReminderFormModel[];
  labels: ReminderLabel[];
}

export interface ReminderFormModel {
  heading: string;
  note: string;
  caseId: string;
  caseLink: string;
  reminderDate: string;
  caseType: string;
  externalCaseId: string;
  isReminder?: boolean;
  id?: string;
  created?: string;
  createdBy?: string;
  modified?: string;
  modifiedBy?: string;
}

export const defaultReminder = {
  heading: '',
  note: '',
  caseId: '',
  caseLink: '',
  reminderDate: '',
  caseType: '',
  externalCaseId: '',
  isReminder: false,
};

export interface ReminderEditFormModel extends ReminderFormModel {
  reminderId: string;
}

export const remindersLabels = [
  { label: 'Ärende', screenReaderOnly: false, sortable: true, shownForStatus: statusCodes.Any },
  { label: 'Senast ändrat', screenReaderOnly: false, sortable: true, shownForStatus: statusCodes.Any },
  { label: 'Skapad av', screenReaderOnly: false, sortable: true, shownForStatus: statusCodes.Any },
  { label: 'Anteckning', screenReaderOnly: false, sortable: true, shownForStatus: statusCodes.Any },
  { label: 'Ärendeknapp', screenReaderOnly: true, sortable: false, shownForStatus: statusCodes.Any },
];

export const emptyReminderList: RemindersData = { reminders: [], labels: remindersLabels };

export const handleReminderResponse: (res: ApiResponse<ReminderResponse[]>) => ReminderFormModel[] = (res) =>
  res.data.map((n: ReminderResponse, idx) => ({
    heading: n.action,
    note: n.note,
    caseId: n.caseId,
    caseLink: n.caseLink,
    reminderDate: n.reminderDate,
    caseType: n.caseType,
    externalCaseId: n.externalCaseId,
    isReminder: true,
    id: n.reminderId,
    created: dayjs(n.created).format('YYYY-MM-DD HH:MM'),
    createdBy: n.createdBy,
    modified: dayjs(n.modified).format('YYYY-MM-DD HH:MM'),
    modifiedBy: n.modifiedBy,
  }));

export interface Action {
  id: number;
  label: string;
  enabled: boolean;
}

export const actions: Action[] = [
  { id: 1, label: 'Följa upp', enabled: true },
  { id: 2, label: 'Komplettera', enabled: true },
  { id: 3, label: 'Förnya tillstånd', enabled: true },
  { id: 4, label: 'Skicka nytt ärende', enabled: true },
  { id: 5, label: 'Övrigt', enabled: true },
];

export const ongoingCasesActions: Action[] = [
  { id: 1, label: 'Följa upp', enabled: true },
  { id: 2, label: 'Komplettera', enabled: true },
  { id: 3, label: 'Övrigt', enabled: true },
];

export const closedCasesActions: Action[] = [
  { id: 1, label: 'Följa upp', enabled: true },
  { id: 2, label: 'Förnya tillstånd', enabled: true },
  { id: 3, label: 'Skicka nytt ärende', enabled: true },
  { id: 4, label: 'Övrigt', enabled: true },
];

export const getReminders: () => Promise<RemindersData> = () =>
  apiService
    .get<ApiResponse<ReminderResponse[]>>('reminders')
    .then((res) => ({ reminders: handleReminderResponse(res.data), labels: remindersLabels } as RemindersData))
    .catch((e) => ({ ...emptyReminderList, error: e.response?.status ?? 'UNKNOWN ERROR' } as RemindersData));

export const handleSendReminder: (reminder: ReminderFormModel) => SendReminder = (reminder) => ({
  action: reminder.heading,
  note: reminder.note,
  caseId: reminder.caseId,
  caseLink: reminder.caseLink,
  caseType: reminder.caseType,
  externalCaseId: reminder.externalCaseId ? reminder.externalCaseId.toString() : reminder.externalCaseId,
  reminderDate: reminder.reminderDate,
});

export const saveReminder: (reminder: ReminderFormModel) => Promise<boolean> = (reminder) => {
  return apiService
    .post('reminders', handleSendReminder(reminder))
    .then(() => Promise.resolve(true))
    .catch((e) => Promise.resolve(false));
};

export const editReminder: (reminder: ReminderFormModel) => Promise<boolean> = (reminder) => {
  return apiService
    .patch(`reminders/${reminder.id}`, handleSendReminder(reminder))
    .then(() => Promise.resolve(true))
    .catch((e) => Promise.resolve(false));
};

export const deleteReminder: (id: string) => Promise<boolean> = (id) => {
  return apiService
    .delete(`reminders/${id}`)
    .then(() => Promise.resolve(true))
    .catch((e) => Promise.resolve(false));
};
