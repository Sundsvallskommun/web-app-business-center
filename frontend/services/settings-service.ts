import { ApiResponse, apiService, Data } from './api-service';

export enum FeedbackLifespan {
  'untilRemoved' = 'untilRemoved',
  'twoWeeks' = 'twoWeeks',
  'oneMonth' = 'oneMonth',
}

export interface ContactChannel {
  contactMethod: string;
  destination: string;
  sendFeedback: boolean;
  alias: string;
}

export const defaultChannel: ContactChannel = {
  contactMethod: '',
  destination: '',
  sendFeedback: true,
  alias: '',
};

export type ContactSettingsResponse = Array<{
  id: string;
  partyId: string;
  contactChannels: ContactChannel[];
  created: string;
  modified: string;
}>;

export interface ContactSettingsUpdateModel {
  id: string;
  contactChannels: ContactChannel[];
}

export interface OverviewFormModel {
  feedbackLifespan: FeedbackLifespan;
}

export const defaultOverviewsSettings: OverviewFormModel = {
  feedbackLifespan: FeedbackLifespan.oneMonth,
};

export interface Contact {
  alias: string;
  contactMethods: {
    SMS: ContactChannel[];
    EMAIL: ContactChannel[];
  };
}

export const defaultContact = {
  alias: '',
  contactMethods: {
    SMS: [],
    EMAIL: [],
  },
};

export interface ContactFormModel {
  id: string;
  contacts: Contact[];
}

export interface FeedbackData extends Data {
  settings: ContactFormModel;
}

export const defaultContactSettings: ContactFormModel = {
  id: '',
  contacts: [],
};

export const handleContactResponse: (res: ApiResponse<ContactSettingsResponse>) => ContactFormModel = (res) => ({
  id: res.data[0].id,
  contacts: Object.values(
    res.data.reduce((accumulator, contactSettingsGroup) => {
      contactSettingsGroup.contactChannels.forEach((channel) => {
        if (!accumulator[channel.alias]) {
          accumulator[channel.alias] = {
            alias: channel.alias,
            contactMethods: {
              SMS: [],
              EMAIL: [],
            },
          };
        }
        if (channel['contactMethod'] == 'SMS') {
          accumulator[channel.alias].contactMethods['SMS'].push(channel);
        }
        if (channel['contactMethod'] == 'EMAIL') {
          accumulator[channel.alias].contactMethods['EMAIL'].push(channel);
        }
      });
      return accumulator;
    }, {})
  ),
});

export const getContactSettings: () => Promise<FeedbackData> = () =>
  apiService
    .get<ApiResponse<ContactSettingsResponse>>('contactsettings')
    .then((res) => ({ settings: handleContactResponse(res.data) } as FeedbackData))
    .catch((e) => ({ settings: defaultContactSettings, error: e.response?.status ?? 'UNKNOWN ERROR' } as FeedbackData));

export const makeContactSettingsUpdateModel: (settings: ContactFormModel) => ContactSettingsUpdateModel = (
  settings
) => {
  return {
    id: settings.id,
    contactChannels: [
      ...settings.contacts
        .reduce((acc, curr) => {
          acc.push(curr.contactMethods['SMS']);
          acc.push(curr.contactMethods['EMAIL']);

          return acc;
        }, [])
        .flat(2),
    ],
  };
};

export const newContactSettings: (settings: ContactFormModel) => Promise<boolean> = (settings) => {
  return apiService
    .post('contactsettings', makeContactSettingsUpdateModel(settings))
    .then(() => Promise.resolve(true))
    .catch((e) => Promise.resolve(false));
};

export const saveContactSettings: (settings: ContactFormModel) => Promise<boolean> = (settings) => {
  return apiService
    .patch('contactsettings', makeContactSettingsUpdateModel(settings))
    .then(() => Promise.resolve(true))
    .catch((e) => Promise.resolve(false));
};
