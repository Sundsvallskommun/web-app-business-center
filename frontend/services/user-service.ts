import { FeedbackLifespan, User } from '@interfaces/user';
import dayjs from 'dayjs';
import { ApiResponse, apiService } from './api-service';

export const emptyUser: User = {
  name: '',
  userSettings: {
    feedbackLifespan: FeedbackLifespan.oneMonth,
    readNotificationsClearedDate: dayjs().toISOString(),
  },
};

const emptyUserResponse: ApiResponse<User> = {
  data: emptyUser,
  message: 'none',
};

export interface UserMetaResponse {
  lastLoginTime?: string;
}

const mockUserMeta: UserMetaResponse = {
  lastLoginTime: dayjs().subtract(180, 'day').toISOString(),
};

// TODO Return actual data when available in api
export const getUserMeta: () => Promise<UserMetaResponse> = () => Promise.resolve(mockUserMeta);

const handleSetUserResponse: (res: ApiResponse<User>) => User = (res) => ({
  name: res.data.name,
  userSettings: {
    feedbackLifespan: res.data.userSettings.feedbackLifespan,
    readNotificationsClearedDate: res.data.userSettings.readNotificationsClearedDate,
  },
});

export const getMe: () => Promise<User> = () => {
  return apiService
    .get<ApiResponse<User>>('me')
    .then((res) => handleSetUserResponse(res.data))
    .catch(() => Promise.reject('No user found'));
};

export const saveUserSettings: (settings: any) => Promise<boolean> = (settings) => {
  return apiService
    .patch('settings', settings)
    .then(() => Promise.resolve(true))
    .catch((e) => {
      return Promise.resolve(false);
    });
};
