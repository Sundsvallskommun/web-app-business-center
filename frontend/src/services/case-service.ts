import { CasePdf, CasePdfData, CasesData, ICase } from '../interfaces/case';
import { statusCodes } from '../interfaces/status-codes';
import { apiService, ApiResponse } from './api-service';

interface CaseResponse {
  caseType: string;
  externalCaseId: string;
  id: string;
  status: string;
  lastStatusChange: string;
  firstSubmitted: string;
  isOpenEErrand: boolean;
}

export const emptyCaseList: CasesData = {
  cases: [],
  labels: [],
};

export const ongoingCasesLabels = [
  { label: 'Ärende', screenReaderOnly: false, sortable: true, shownForStatus: statusCodes.Any },
  // { label: 'Ansvarig förvaltning', screenReaderOnly: false, sortable: true, shownForStatus: statusCodes.Any },
  { label: 'Senast ändrat', screenReaderOnly: false, sortable: true, shownForStatus: statusCodes.Any },
  { label: 'Status', screenReaderOnly: false, sortable: true, shownForStatus: statusCodes.Any },
  { label: 'Skapa egen påminnelse', screenReaderOnly: false, sortable: false, shownForStatus: statusCodes.Any },
  { label: 'Ärendeknapp', screenReaderOnly: true, sortable: false, shownForStatus: statusCodes.Any },
];

// getCaseStatus i confluence
const statusMap = {
  Inskickat: { code: statusCodes.Ongoing, color: 'info' },
  'Tilldelat för handläggning': { code: statusCodes.Ongoing, color: 'info' },
  'Under behandling': { code: statusCodes.Ongoing, color: 'info' },
  'Komplettering behövs': { code: statusCodes.Ongoing, color: 'warning' },
  'Påminnelse om komplettering': { code: statusCodes.Ongoing, color: 'error' },
  'Komplettering inkommen, behandling fortsätter': {
    code: statusCodes.Ongoing,
    color: 'info',
  },
  'Beslut finns, se separat information': {
    code: statusCodes.Ongoing,
    color: 'info',
  },
  'Väntar på komplettering': { code: statusCodes.Ongoing, color: 'warning' },
  Kompletterad: { code: statusCodes.Ongoing, color: 'info' },
  Klart: { code: statusCodes.Approved, color: 'neutral' },
  Avslutat: { code: statusCodes.Approved, color: 'neutral' },
  'Ärendet arkiveras': { code: statusCodes.Rejected, color: 'neutral' },
  'Sparat ärende': { code: statusCodes.Rejected, color: 'neutral' },
};

export const mapStatus = (s: string) => {
  return Object.keys(statusMap).includes(s)
    ? { code: statusMap[s].code, color: statusMap[s].color, label: s }
    : { code: statusCodes.Ongoing, color: 'neutral', label: s };
};

export const handleCase = (n: CaseResponse) => ({
  externalCaseId: n.externalCaseId,
  caseId: n.id,
  subject: {
    caseType: n.caseType,
    meta: {
      created: n.firstSubmitted,
      modified: n.lastStatusChange,
    },
  },
  department: '--',
  validFrom: '--',
  validTo: '--',
  serviceDate: '--',
  status: mapStatus(n.status),
  lastStatusChange: n.lastStatusChange,
});

export const handleCaseResponse: (data: CaseResponse[]) => ICase[] = (data) => data.map(handleCase);

export const casesHandler = (data) => ({
  cases: handleCaseResponse(data),
  labels: [],
});

export const getOngoing: (cs: CasesData) => CasesData = (cs) => ({
  ...cs,
  labels: ongoingCasesLabels,
  cases: cs.cases.filter((c) => c.status.code === statusCodes.Ongoing),
});

export const closedCasesLabels = [
  { label: 'Ärende', screenReaderOnly: false, sortable: true, shownForStatus: statusCodes.Any },
  { label: 'Senast ändrat', screenReaderOnly: false, sortable: true, shownForStatus: statusCodes.Any },
  { label: 'Status', screenReaderOnly: false, sortable: true, shownForStatus: statusCodes.Any },
  { label: 'Skapa egen påminnelse', screenReaderOnly: false, sortable: false, shownForStatus: statusCodes.Any },
  { label: 'Ärendeknapp', screenReaderOnly: true, sortable: false, shownForStatus: statusCodes.Any },
];

export const getClosed: (cs: CasesData) => CasesData = (cs) => ({
  ...cs,
  labels: closedCasesLabels,
  cases: cs.cases.filter((c) => c.status.code === statusCodes.Rejected || c.status.code === statusCodes.Approved),
});

export const getCasesInNeedOfData: (cs: CasesData) => CasesData = (cs) => ({
  ...cs,
  labels: ongoingCasesLabels,
  cases: cs.cases.filter((c) =>
    ['Väntar på komplettering', 'Påminnelse om komplettering', 'Komplettering behövs'].includes(c.status.label)
  ),
});

export const getCasePdf: (caseId: string) => Promise<CasePdfData> = (caseId) =>
  apiService
    .get<ApiResponse<CasePdf>>(`casepdf/${caseId}`)
    .then((res) => ({ pdf: res.data.data }))
    .catch(
      (e) => ({ pdf: { externalCaseId: '', base64: '' }, error: e.response?.status ?? 'UNKNOWN ERROR' }) as CasePdfData
    );
