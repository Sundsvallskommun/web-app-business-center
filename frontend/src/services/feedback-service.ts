import { apiService } from './api-service';

interface FeedbackModal {
  body: string;
}

export const sendFeedback: (feedback: FeedbackModal) => Promise<boolean> = (feedback) => {
  return apiService
    .post('feedback', feedback)
    .then(() => Promise.resolve(true))
    .catch((e) => Promise.resolve(false));
};
