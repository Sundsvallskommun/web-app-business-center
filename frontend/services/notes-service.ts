import dayjs from 'dayjs';
import { statusCodes } from '../interfaces/status-codes';
import { ApiResponse, apiService, Data } from './api-service';
import { ReminderFormModel } from './reminder-service';
import { ApiResponseMeta } from '@interfaces/service';

export interface NoteResponse {
  id: string;
  subject: string;
  body: string;
  caseId: string;
  caseLink: string;
  caseType: string;
  externalCaseId: string;
  createdBy?: string;
  created: string;
  modified: string;
  modifiedBy: string;
  context: string;
  role: string;
  clientId: string;
}

const emptyNotesResponse: ApiResponse<{ notes: NoteResponse[] }> = {
  data: { notes: [] },
  message: 'none',
};

export type SendNoteOmittedTypes = 'id' | 'partyId' | 'created' | 'modified' | 'modifiedBy';

export interface SendNote {
  subject: string;
  body: string;
  caseId: string;
  caseLink: string;
  caseType: string;
  externalCaseId: string;
  createdBy?: string;
}

export interface NoteLabel {
  label: string;
  sortable: boolean;
  shownForStatus: statusCodes;
  screenReaderOnly: boolean;
}

export interface NotesData extends Data {
  _meta?: ApiResponseMeta;
  notes: NoteResponse[];
  labels: NoteLabel[];
}

export const noteLabels: NoteLabel[] = [
  { label: 'Ärende', sortable: false, screenReaderOnly: true, shownForStatus: statusCodes.Any },
  { label: 'Innehåll', sortable: false, screenReaderOnly: true, shownForStatus: statusCodes.Any },
];

export const emptyNotesList: NotesData = { notes: [], labels: noteLabels };

export const handleNotesResponse: (res: ApiResponse<{ notes: NoteResponse[] }>) => ReminderFormModel[] = (res) =>
  res.data.notes.map((n: NoteResponse) => ({
    heading: n.subject,
    note: n.body,
    caseId: n.caseId,
    caseLink: n.caseLink,
    reminderDate: '',
    caseType: n.caseType,
    externalCaseId: n.externalCaseId,
    isReminder: false,
    id: n.id,
    created: dayjs(n.created).format('YYYY-MM-DD HH:MM'),
    createdBy: n.createdBy,
    modified: dayjs(n.modified).format('YYYY-MM-DD HH:MM'),
    modifiedBy: n.modifiedBy,
  }));

export const getNotes: () => Promise<NotesData> = () =>
  apiService
    .get<ApiResponse<{ notes: NoteResponse[]; _meta: ApiResponseMeta }>>('notes')
    .then(
      (res) =>
        ({
          notes: handleNotesResponse(res.data),
          labels: noteLabels,
          _meta: res.data.data._meta,
        } as unknown as NotesData)
    )
    .catch(
      (e) =>
        ({
          notes: handleNotesResponse(emptyNotesResponse),
          labels: noteLabels,
          error: e.response?.status ?? 'UNKNOWN ERROR',
        } as unknown as NotesData)
    );

export const handleSendNote: (reminder: ReminderFormModel) => SendNote = (reminder) => ({
  subject: reminder.heading,
  body: reminder.note,
  caseId: reminder.caseId,
  caseLink: reminder.caseLink,
  caseType: reminder.caseType,
  externalCaseId: reminder.externalCaseId,
});

export const saveNote: (note: ReminderFormModel) => Promise<boolean> = (note) => {
  return apiService
    .post('notes', handleSendNote(note))
    .then(() => Promise.resolve(true))
    .catch((e) => Promise.resolve(false));
};

export const editNote: (note: ReminderFormModel) => Promise<boolean> = (note) => {
  return apiService
    .patch(`notes/${note.id}`, handleSendNote(note))
    .then(() => Promise.resolve(true))
    .catch((e) => Promise.resolve(false));
};

export const deleteNote: (id: string) => Promise<boolean> = (id) => {
  return apiService
    .delete(`notes/${id}`)
    .then(() => Promise.resolve(true))
    .catch((e) => Promise.resolve(false));
};
