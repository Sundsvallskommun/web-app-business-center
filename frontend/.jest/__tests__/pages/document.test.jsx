import '@jestRoot/__mocks__/services.mock';
import '@jestRoot/__mocks__/context.mock';
import ShallowRenderer from 'react-test-renderer/shallow';
import MyDocument from '@pages/_document';

describe('MyDocument', () => {
  const renderer = new ShallowRenderer();
  renderer.render(<MyDocument />);
  const result = renderer.getRenderOutput();

  it('renders MyDocument unchanged', () => {
    expect(result).toMatchSnapshot();
  });
});
