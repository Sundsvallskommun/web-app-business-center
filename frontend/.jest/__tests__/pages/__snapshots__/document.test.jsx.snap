// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyDocument renders MyDocument unchanged 1`] = `
<Html
  lang="sv"
>
  <Head />
  <body>
    <Main />
    <NextScript />
  </body>
</Html>
`;
